pycalphad/codegen/phase_record_factory.py

```diff
@@ -0,0 +1,78 @@
+import pycalphad.variables as v
+from pycalphad.codegen.sympydiff_utils import build_functions
+from pycalphad.core.utils import get_pure_elements, unpack_components, \
+    extract_parameters, get_state_variables
+from pycalphad.core.phase_rec import PhaseRecord
+from pycalphad.core.constraints import build_constraints
+from itertools import repeat
+from functools import lru_cache
+import numpy as np
+
+class PhaseRecordFactory(object):
+    def __init__(self, dbf, comps, state_variables, models, parameters=None):
+        self.comps = sorted(unpack_components(dbf, comps))
+        self.pure_elements = get_pure_elements(dbf, comps)
+        self.nonvacant_elements = sorted([x for x in self.pure_elements if x != 'VA'])
+        self.molar_masses = np.array([dbf.refstates[x]['mass'] for x in self.nonvacant_elements], dtype='float')
+        parameters = parameters if parameters is not None else {}
+        self.models = models
+        self.state_variables = sorted(get_state_variables(models=models, conds=state_variables), key=str)
+        self.param_symbols, self.param_values = extract_parameters(parameters)
+
+        if len(self.param_values.shape) > 1:
+            self.param_values = self.param_values[0]
+
+    def update_parameters(self, parameters):
+        new_param_symbols, new_param_values = extract_parameters(parameters)
+        if len(new_param_values.shape) > 1:
+            new_param_values = new_param_values[0]
```

Comment:
It just takes the first set of parameters so it can be initialized with a valid parameter array. The caller is supposed to call `PhaseRecordFactory.update_parameters()` to iterate through a set of parameters in a loop, or use the `PhaseRecord.{obj,prop}_parameters_2d()` API to get a vectorized result based on the set of parameters. In both cases, the caller should be managing that state.